File: 0785-is-graph-bipartite/0785-is-graph-bipartite.java
class Solution {
    
    private int n;
    private int[] arr;
    
    private boolean DFS(int node, boolean color, int[][] adj){
        
        if(arr[node] != -1)
            return arr[node] == (color ? 1 : 0);
        
        arr[node] = color ? 1 : 0; // Convert boolean color to int
        boolean ans = true;
        
        for(int neigh : adj[node]){
            ans &=  DFS(neigh, !color, adj);
            
            if(!ans)
                break;
        }
        
        return ans;
    }
    
    public boolean isBipartite(int[][] adj) {
        
        this.n = adj.length;
        this.arr = new int[n];
        
        for(int i = 0; i < n; i++)
            arr[i] = -1;
        
        boolean ans = true;
        
        for(int i=0;i<n;i++){
            if(arr[i]==-1)
                ans &= DFS(i,false,adj);
            
            if(!ans)
                break;
                
        }
        
        return ans;
    }
}

File: 2025-and-before/0785-is-graph-bipartite/0785-is-graph-bipartite.cpp
class Solution {
public:
    bool DFS(int curr,int col, vector<int>&color, vector<vector<int>>&adj)
    {
        color[curr]=col;
        bool ans=true;
        for(int i : adj[curr])
        {
            if(color[i]!=-1 and color[i]==col)
                ans= ans & false;
            
            else if(color[i]==-1)
                ans =  ans &  DFS(i,!col,color,adj);
            
            if(!ans)
                return false;
        }
        
        return  true;
        
    }
    
    bool isBipartite(vector<vector<int>>& graph) {
        int n=graph.size();
        vector<int>color(n,-1);
        
        for(int i=0;i<n;i++)
        {
            if(color[i]==-1 and DFS(i,0,color,graph)==false)
                return false;
        }
        
        return true;      
        
        
    }
};